Dsikora369_code/task_1/titanic.py

```diff
@@ -0,0 +1,286 @@
+
+Sibsp and Parch refer to family, so it is logical to replace them with one column and create 'Alone' column for people without family members.
+"""
+
+train['FamilySize'] = train['SibSp'] + train['Parch'] + 1
+train[['FamilySize', 'Survived']].groupby(['FamilySize'],
+                                          as_index=False).mean()
+
+train['Alone'] = 0
+train.loc[train['FamilySize'] == 1, 'Alone'] = 1
+train[['Alone', 'Survived']].groupby(['Alone'], as_index=False).mean()
+
+"""'Fare' has NaNs, so they will be replaced to mean. Also, for future model it is better to replace 'Fare' with its categorical representation. """
+
+train['Fare'] = train['Fare'].fillna(train['Fare'].median())
+train['CategoricalFare'] = pd.qcut(train['Fare'], 5)
+train[['CategoricalFare', 'Survived']].groupby(
+    ['CategoricalFare'], as_index=False).mean()
+
+"""Also, it is very useful to notice whether passanger had a cabin or not and which cabin.
+
+"""
+
+
+def get_cabin(cabin):
+    for i in str(cabin):
+        if i in ['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'Z']:
+            return i
+
+
+train['Cabin'] = train['Cabin'].fillna('Z')
+train['BL'] = train['Cabin'].apply(get_cabin)
+train[['BL', 'Survived']].groupby(['BL'], as_index=False).mean()
+
+"""In 'Age' we will replace NaNs with mean random and transform to categorical."""
+
+age_avg = train['Age'].mean()
+age_std = train['Age'].std()
+age_null_count = train['Age'].isnull().sum()
+age_null_random_list = np.random.randint(age_avg - age_std, age_avg + age_std,
+                                         size=age_null_count)
+train['Age'][np.isnan(train['Age'])] = age_null_random_list
+train['Age'] = train['Age'].astype(int)
+train['CategoricalAge'] = pd.cut(train['Age'], 5)
+train[['CategoricalAge', 'Survived']].groupby(['CategoricalAge'],
+                                              as_index=False).mean()
+
+"""We can find out different titles of passangers."""
+
+
+def get_title(name):
+    title_search = re.search(r'([A-Za-z]+)\.', name)
+    if title_search:
+        return title_search.group(1)
+    return ""
+
+
+train['Title'] = train['Name'].apply(get_title)
+pd.crosstab(train['Title'], train['Sex'])
+
+"""So we can transform titles to categorical."""
+
+train['Title'] = train['Title'].replace(['Lady', 'Countess', 'Capt', 'Col',
+                                         'Don', 'Dr', 'Major', 'Rev', 'Sir',
+                                         'Jonkheer', 'Dona'], 'Rare')
+
+train['Title'] = train['Title'].replace('Mlle', 'Miss')
+train['Title'] = train['Title'].replace('Ms', 'Miss')
+train['Title'] = train['Title'].replace('Mme', 'Mrs')
+
+train[['Title', 'Survived']].groupby(['Title'], as_index=False).mean()
+
+"""Column 'Embarked' has NaNs that will be replced with median."""
+
+train['Embarked'] = train['Embarked'].fillna(train['Embarked'].mode()[0])
+
+"""Adding Ageclass feature.
+
+"""
+
+train['Age*Class'] = train.Age * train.Pclass

```

Comment:
You may consider doing such a transformation automatically to many features using `PolynomialFeatures`
https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.PolynomialFeatures.html